File: tree/110.c
/**
 * dfs use -1 to show when the node is unbalanced, then go recursive 
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

int checkBalance(struct TreeNode* root)
{
    if(root==NULL) return 0;
    int leftheight = checkBalance(root->left);
    if(leftheight == -1) return -1;
    int rightheight = checkBalance(root->right);
    if(rightheight == -1) return -1;
    if(abs(leftheight - rightheight) > 1) return -1;
    return leftheight > rightheight ? leftheight + 1 : rightheight + 1;
}
bool isBalanced(struct TreeNode* root) {
    return checkBalance(root) != -1;
}

File: tree/110.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isBalanced(root *TreeNode) bool {
    if root == nil {
        return true
    }
    var leftH, rightH int
    height(root.Left, 0, &leftH)
    height(root.Right, 0, &rightH)
    if abs(leftH, rightH) > 1 {
        return false
    }else {
        return isBalanced(root.Left) && isBalanced(root.Right)
    }
}

func abs(a int, b int) int {
    if a < b {
        return b - a 
    }else {
        return a - b
    }
}

func height(root *TreeNode, cur int, ret *int) {
    if root == nil {
        return
    }
    if root.Left == nil && root.Right == nil {
        if cur+1 > *ret {
            *ret = cur+1
        }
        return
    }
    if root.Left != nil {
        height(root.Left, cur+1, ret)
    }
    if root.Right != nil {
        height(root.Right, cur+1, ret)
    }
}
